ori/volume/server/volume_create.go

```diff
@@ -17,260 +17,167 @@ package server
-				DataSourceRef: &corev1.TypedObjectReference{
-					APIGroup: pointer.String(storagev1alpha1.SchemeGroupVersion.String()),
-					Kind:     "Volume",
-					Name:     pvc.Name,
-				},
-			},
-		}
-		metautils.SetLabel(imagePvc, volumev1alpha1.ManagerLabel, volumev1alpha1.VolumeCephletManager)
+	if err := s.validateVolumeClass(oriVolume); err != nil {
+		return fmt.Errorf("invalid volume class: %w", err)
+	}
+	log.V(2).Info("Validated class")
 
+	validatedImage, err := s.validateImage(ctx, log, oriVolume)
+	if err != nil {
+		return fmt.Errorf("invalid volume image: %w", err)
 	}
+	log.V(2).Info("Validated image")
 
-	if err := apiutils.SetObjectMetadata(pvc, volume.Metadata); err != nil {
-		return nil, err
+	if err := s.validateSize(oriVolume, validatedImage); err != nil {
+		return fmt.Errorf("invalid volume size: %w", err)
 	}
-	metautils.SetLabel(pvc, volumev1alpha1.ManagerLabel, volumev1alpha1.VolumeCephletManager)
-	apiutils.SetVolumeClassLabel(pvc, volume.Spec.Class)
+	log.V(2).Info("Validated size")
+
+	return nil
 
-	return &AggregateCephVolume{
-		Pvc:         pvc,
-		ImagePvc:    imagePvc,
-		VolumeClass: volumeClass,
-	}, nil
 }
 
-func (s *Server) handleImagePopulation(ctx context.Context, log logr.Logger, volume *AggregateCephVolume) error {
-	if volume.ImagePvc == nil {
-		return nil
-	}
+func (s *Server) prepareOSImage(ctx context.Context, log logr.Logger, osImageName string) (osImageId string, retErr error) {
+	c, cleanup := setupCleaner(ctx, log, &retErr)
+	defer cleanup()
 
-	snapshot := &snapshotv1.VolumeSnapshot{}
-	if err := s.client.Get(ctx, client.ObjectKeyFromObject(volume.ImagePvc), snapshot); err != nil {
-		if !errors.IsNotFound(err) {
-			return fmt.Errorf("unable to get snapshot: %w", err)
-		}
-		log.V(1).Info("Requested snapshot not found: create image pvc and snapshot it.")
-		return s.createSnapshot(ctx, log, volume)
+	osImageId, found, err := s.provisioner.GetOsImage(ctx, osImageName)
+	if err != nil {
+		return "", fmt.Errorf("unable to get os image: %w", err)
 	}
 
-	if !pointer.BoolDeref(snapshot.Status.ReadyToUse, false) || snapshot.Status.RestoreSize == nil {
-		log.Info("Referenced snapshot not ready or RestoreSize not defined", "snapshotName", snapshot.Name)
-		return nil
+	if found {
+		return osImageId, nil
 	}
 
-	volumeSizeBytes := volume.Pvc.Spec.Resources.Requests.Storage().Value()
-	if volumeSizeBytes < snapshot.Status.RestoreSize.Value() {
-		log.Info(fmt.Sprintf("Requested volume size %d is less than the size %d for the source snapshot", volumeSizeBytes, snapshot.Status.RestoreSize.Value()), "snapshotName", snapshot.Name)
-		//r.Eventf(volume, corev1.EventTypeWarning, ReasonVolumeSizeToSmall, "Requested volume size %d is less than the size %d for the source snapshot %s", volumeSizeBytes, snapshot.Status.RestoreSize.Value(), snapshot.Name)
+	osImageId = s.idGen.Generate()
+	if err := s.provisioner.CreateOsImage(ctx, osImageName, osImageId); err != nil {
```

Comment:
Why do we need to generate an id for an os image? The unique identifier of an image should be its digest (coming from `Image.Digest()`). The downloading entity should be able to pull an image by the image name and return the digest, additional mappings limit re-use / make the whole flow more complex.